src/object.d

```diff
@@ -2922,21 +2918,53 @@ unittest
     assert(postblitRecurseOrder == order);
 }
 
+// destroy
 /********
-Destroys the given object and sets it back to its initial state. It's used to
-_destroy an object, calling its destructor or finalizer so it no longer
-references any other objects. It does $(I not) initiate a GC cycle or free
-any GC memory.
+Destroys an object by means of calling its destructor, if present. For
+types `T` that define a destructor, `destroy` then also fills the
+object with `T.init` to ensure the object's destructor can be called
+again (e.g. when the object goes out of scope). Unlike $(LREF
+__delete), `destroy` does not release memory or other resources by
+itself (though it is possible that destructors it calls do so).
+
+Note: A `struct` type may have a destructor even if it does not
+directly define `~this()`, if at least one field has a destructor
+(transitively). A `class` type may have a destructor even if it does
+not directly define `~this()`, if at least one field has a destructor
+(transitively), or if the destructor is inherited.
+
+Detailed behavior of `destroy(obj)` depends on the type `T` as
+follows:
+
+$(UL $(LI If `T` is a primitive type, a `struct` types that does not
+have a destructor, or a `class` type that does not have a destructor,
+the call to has no effect.)
+
+$(LI For `class` and `interface` types, the call `destroy(obj)` has no
+effect if `obj` is `null`.)
+
+$(LI For `struct` and `class` types that have a destructor (either
+defined explicitly or synthesized by the compiler), the destructor is
+called. Then the object is initialized using
+`T.init`. Reinitialization occurs even if the type disables default
+construction with `@disable this()`.)
+
+$(LI For `interface` types, the dynamic `class` object is fetched from
+the interface object, and then `destroy` is applied to it.)
+
+$(LI For static and dynamic arrays, `destroy` is applied right-to-left
+to all elements. This means multidimensional arrays are destroyed in
+depth, transitively (behavior different from $(LREF __delete), which
+is by necessity shallow). If one or more destructors throw an
+exception, the calls are completed and then the first exception
+encountered is rethrown.))
 */
-void destroy(T)(T obj) if (is(T == class))
+void destroy(T)(T obj) if (is(T == class) || (is(T == interface)))
```

Comment:
I think there is a pre-existing bug in here, as I think the static test will pass for `extern(C++)` classes, which is definitely NOT correct to call `rt_finalize` on. Something to think about as we move towards betterC.

Reply:
And Objective-C classes. There’s a recent change where `__traits(getLinkage)` supports classes.